working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+can understand the syntax. Grammatical context is lowest&mdash;they literally
+just need to look at the enclosing text. Lexical scoping is next. They have
+to look at surrounding block scopes and know what's imported. Static types are
+the last&mdash;they have to essentially run the type inference engine in their
+head.
+
+In this proposal, we take approach 2: we use the grammatical context. I think
+this makes the right trade-off between familiar syntax and relatively local
+reasoning, at the expense of a slightly more complex somewhat redundant set of
+patterns. This approach was [directly inspired by Swift][swift], so there's
+evidence this is a good balance.
+
+[swift]: https://docs.swift.org/swift-book/ReferenceManual/Patterns.html
+
+## Two kinds of patterns
+
+Patterns generally do one of three things:
+
+1.  Perform some kind of test on an object to decide whether to execute some
+    piece of code or not. In other words, if the object **matches** the pattern,
+    then evaluate the "body" of some surrounding construct. If the object does
+    not match the value, the we say the pattern was **refuted**.
+
+2.  **Bind** a new variable to the value.
+
+3.  Extract data from some object by **destructuring** it and applying
+    subpatterns to the extracted data.
+
+Functional languages lump those three together into a single grammar, "pattern",
+and allow them to be freely mixed. Instead, for Dart, we have two separate
+categories: **matchers** and **binders**. Matchers are all of the refutable
+patterns that can fail to match a value. Binders are the other patterns that can
+never fail and exist to bind new variables.
+
+Destructuring patterns like list and map patterns that contain subpatterns
+exist in both categories. There is a list binder and a list matcher. The only
+difference is that the former contains binders as subpatterns and the latter
+matchers. (This is where the redundancy comes in.)
+
+Places in the language where patterns can be embedded use the category that
+makes sense for that context. For example, if we allow patterns in variable
+declarations, like:
+
+```dart
+var [i, j] = [1, 2];
+```
+
+Then the pattern appearing after `var` is a binder, because that's the intent of
+a variable declaration. (In this example, it's a list binder containing two
+variable binders.) On the other hand, in some kind of match statement, the
+pattern is a matcher because the main purpose of the statement is control flow:
+
+```dart
+match (value) {
+  case [1, 2] => print("matched");
+}
+```
+
+(Here we have a list *matcher* containing two literal matchers.)
+
+### Disambiguating
+
+The two categories are useful because some patterns do not appear in both
+categories. There is a pattern to determine if a value is equal to a given named
+constant, but that is only a matcher. Likewise, the pattern to bind a new
+variable is only a binder. In this way, a bare identifier is unambiguous: in a
```

Comment:
> Does it mean that in a binder, you write `x` to bind a match, and in a matcher you have to write `var x`?

Yes.

> Can you write `var x` in a binder?

No, because `var` only comes from `bindMatcher`, which isn't a binder.

> because otherwise a tuple binding like `(int x, int y) = pair;` would need to be `(int, int) (x, y) = pair;` or `var (int x, int y) = pair;`, neither of which are as convenient.

In most cases, you don't need to type the elements because they can be inferred from the initializer, so it's just:

```dart
var (x, y) = pair;
```

If you do want to type them (maybe to upcast or because you don't have an initializer), then it would be the latter:

```dart
var (int x, int y) = pair;
```

Not as convenience, I guess, but I think having a leading declaration keyword is a good thing. :)